src/networking.c

```diff
@@ -3178,7 +3286,18 @@ void readQueryFromClient(connection *conn) {
     client *c = connGetPrivateData(conn);
     int nread, big_arg = 0;
     size_t qblen, readlen;
-    if (!(c->io_flags & CLIENT_IO_READ_ENABLED)) return;
+
+    if (!(c->io_flags & CLIENT_IO_READ_ENABLED)) {
+        /* Replica client can only receive ACK messages. Make sure to notify
+         * the main thread that it needs to return back to IO thread to read
+         * the message.
+         * Note this is not an active notification - main thread checks if the
+         * flag is raised during beforeSleep - see sendReplicasToIOThread */
+        if (c->flags & CLIENT_SLAVE && c->running_tid == IOTHREAD_MAIN_THREAD_ID)
+            atomicSetWithSync(c->pending_ack, 1);
+        return;
+    }
+
```

Comment:
here we can not read c->flags, it is a data race. and i think `c->pending_ack` can be more generic, maybe `pending_read`.

```suggestion
    if (!(c->io_flags & CLIENT_IO_READ_ENABLED)) {
        atomicSetWithSync(c->pending_ack, 1);
        return;
    } else {
        atomicSetWithSync(c->pending_ack, 0);
    }
```

Reply:
Ah yes, TSan did not report it but indeed it has data race. I agree on the `pending_read` change, but I think `atomicSetWithSync(c->pending_ack, 0);` is bad here as in the context of no IO threads this is unnecessary atomic operation. 

EDIT: Can't think of a better place to put it though. So I think best to check
`else if (server.io_threads_num > 1) ...`

Reply:
> else if (server.io_threads_num > 1) ...

looks good to me